apps/heft/src/cli/HeftActionRunner.ts

```diff
@@ -0,0 +1,685 @@
+            isForbidden = true;
+            break;
+          }
+        }
+        if (isForbidden) {
+          // If it's forbidden and we're ignoring it, remove from the unseenFilePaths set, since we
+          // don't want to add it to the seenSourceFilePaths set below.
+          if (ignoreForbidden) {
+            unseenFilePaths.delete(filePath);
+          } else {
+            throw new Error(`Cannot change the file at path "${filePath}" while running watch mode.`);
+          }
+        } else {
+          seenFilePaths.add(filePath);
+        }
+      }
+
+      // Determine which files are ignored or otherwise and stash them away for later.
+      // We can perform this check in one call to git to save time.
+      const unseenIgnoredFilePaths: Set<string> = await git.checkIgnore(unseenFilePaths);
+      const unseenSourceFilePaths: Set<string> = Selection.difference(
+        unseenFilePaths,
+        unseenIgnoredFilePaths
+      );
+      for (const sourceFilePath of unseenSourceFilePaths) {
+        seenSourceFilePaths.add(sourceFilePath);
+      }
+    }
+  }
+
+  function generateChangeHash(filePath: string, fileStats?: fs.Stats): string {
+    // watcher.options.alwaysStat is true, so we can use the stats object directly.
+    // It should only be undefined when the file has been deleted.
+    if (fileStats) {
+      // Base the hash on the modification time, change time, size, and path
+      return crypto
+        .createHash('sha1')
+        .update(filePath)
+        .update(fileStats.mtimeMs.toString())
+        .update(fileStats.ctimeMs.toString())
+        .update(fileStats.size.toString())
+        .digest('hex');
+    } else {
+      return REMOVED_CHANGE_STATE;
+    }
+  }
+
+  function generateChangeState(filePath: string, stats?: fs.Stats): IChangedFileState {
+    const version: string = generateChangeHash(filePath, stats);
+    const isSourceFile: boolean = seenSourceFilePaths.has(filePath);
+    return { isSourceFile, version };
+  }
+
+  function onChange(relativeFilePath: string, fileStats?: fs.Stats): void {
+    // watcher.options.cwd is set below, use to resolve the absolute path
+    const filePath: string = path.join(watcher.options.cwd!, relativeFilePath);
+    changedFileStats.set(filePath, fileStats);
+    resolveFileChange();
+  }
+
+  function createFileChangePromise(): Promise<void> {
+    return new Promise((resolve: () => void, reject: (error: Error) => void) => {
+      resolveFileChange = resolve;
+      rejectFileChange = reject;
+    });
+  }
+
+  // Before we enter the main loop, hydrate initial state and yield the changes.
+  const initialFilePaths: Set<string> = new Set();
+  for (const [directory, filenames] of Object.entries(watcher.getWatched())) {
+    for (const filename of filenames) {
+      const filePath: string = path.resolve(watcher.options.cwd!, directory, filename);
+      if (Path.isUnder(filePath, watcher.options.cwd!)) {
```

Comment:
Same question about perf. Removing use of this API was a huge part of the gitRepoState performance improvement.